### PR TITLE
Add TLS Support

### PR DESCRIPTION
**Pull request by niedhui**
UCP #87 

1. add mTLS among TiDB components #174 
~~2. add TLS support between MySQL client and TiDB #177~~

to enable mTLS, add extra CLI arguments
```
--cluster-ca=/path/ca.pem  --cluster-cert=/path/client.pem --cluster-key=/path/client-key.pem 
```

#### Question
~~Do we need to support two certificates like TiDB server, to support enable cluster mTLS and database TSL separately?~~
~~If so, we need to apply six CLI argument to enable both~~
```
--ssl-ca=/path/ca.pem --ssl-cert=/path/cert.pem --ssl-key=/path/key.pem \
  --cluster-ssl-ca=/path/ca.pem --cluster-ssl-cert=/path/cert.pem --cluster-ssl-key=/path/key.pem
```

### CODE REVIEW

**Comment by you06** [comment]
The score has been corrected, it's 1680 now.

File: cmd/tidb-dashboard/main.go [hunk 1]
```diff
@@ -80,11 +82,28 @@ func NewCLIConfig() *DashboardCLIConfig {
 	flag.Int64Var(&cfg.KVFileStartTime, "keyviz-file-start", 0, "(debug) start time for file range in file mode")
 	flag.Int64Var(&cfg.KVFileEndTime, "keyviz-file-end", 0, "(debug) end time for file range in file mode")
 
+	caPath := flag.String("cacert", "", "path of file that contains list of trusted SSL CAs.")
```
Comment (breeswish) [inline cmd/tidb-dashboard/main.go]: maybe we can call it `cluster-ca`

File: cmd/tidb-dashboard/main.go [hunk 2]
```diff
@@ -80,11 +82,28 @@ func NewCLIConfig() *DashboardCLIConfig {
...
 
+	if len(*caPath) != 0 && len(*certPath) != 0 && len(*keyPath) != 0 {
+		tlsInfo := transport.TLSInfo{
+			CertFile:      *certPath,
+			KeyFile:       *keyPath,
+			TrustedCAFile: *caPath,
+		}
+		tlsConfig, err := tlsInfo.ClientConfig()
+		if err != nil {
+			fmt.Println("failed to load certificates", err)
```
[the remark this answers is not in the document]
Comment (niedhui) [reply cmd/tidb-dashboard/main.go]: done

File: pkg/apiserver/profiling/router.go [hunk 3]
```diff
@@ -43,7 +46,16 @@ func NewService(config *config.Config, db *dbstore.DB) *Service {
 	if err != nil {
 		log.Fatal("Failed to initialize database", zap.Error(err))
 	}
-	return &Service{config: config, db: db, tasks: sync.Map{}}
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialTLS: func(network, addr string) (net.Conn, error) {
```
[the remark this answers is not in the document]
Comment (niedhui) [reply pkg/apiserver/profiling/router.go]: right 😄

File: cmd/tidb-dashboard/main.go (hunk 1 above)
Comment (niedhui) [reply cmd/tidb-dashboard/main.go]: Change to `cluster-ca`, `cluster-cert` and `cluster-key`

File: cmd/tidb-dashboard/main.go [hunk 4]
```diff
@@ -147,7 +166,7 @@ func main() {
 	remoteDataProvider := &keyvisualregion.PDDataProvider{
 		FileStartTime:  cliConfig.KVFileStartTime,
 		FileEndTime:    cliConfig.KVFileEndTime,
-		PeriodicGetter: keyvisualinput.NewAPIPeriodicGetter(cliConfig.CoreConfig.PDEndPoint),
+		PeriodicGetter: keyvisualinput.NewAPIPeriodicGetter(cliConfig.CoreConfig.PDEndPoint, httpClient),
```
[the remark this answers is not in the document]
Comment (niedhui) [reply cmd/tidb-dashboard/main.go]: , I had added some normalize logic to leave the command argument unchanged, and keep code less changed. The `PDEndPoint` of `Config` still contains the protocol part. Maybe we can unify all the internal request related things in another PR

**Comment by breeswish** [comment]
I finally successfully visited the Dashboard of a TLS cluster.. Looks like log exporting doesn't work. Other functionalities work fine. Could you take a look at log exporting?

**Comment by niedhui** [comment]
> I finally successfully visited the Dashboard of a TLS cluster.. Looks like log exporting doesn't work. Other functionalities work fine. Could you take a look at log exporting?

Missing the `grpc` part. It should work now.

**Comment by niedhui** [comment]
> Mostly LGTM! I figure out that the [dashboard initializer](https://github.com/pingcap/tidb-ansible/blob/master/scripts/dashboard_topo.py) in tidb-ansible also does not support TLS. Could you help update it as well?

Sure. Do you know what is the minimal version of python2 we should support in `tidb-ansible`? `2.7.9` up or below? since after `2.7.9`, doing client tls authentication is a litter easier(less code)

**Comment by breeswish** [comment]
@niedhui We officially use CentOS 7, which ships with Python 2.7.5 according to the documentation.

**Comment by breeswish** [comment]
A friendly hint: https://github.com/pingcap-incubator/tidb-dashboard/pull/191 is merged, which contains a `http://` need to be changed :)

**Comment by niedhui** [comment]
> A friendly hint: #191 is merged, which contains a `http://` need to be changed :)

Seems the request is sent to alertmanager, does it support TLS client authentication? or I need to deploy a alertmanager behind an Nginx ?

**Comment by breeswish** [comment]
@niedhui Oh, sorry. Thanks for the remind! I've checked out the code of alert manager, and it does not have TLS now.

**Comment by breeswish** [comment]
Nice!

**Comment by breeswish** [comment]
PR is merged. Feel free to update https://github.com/pingcap/pd/pull/2224 :)

**Pull request merged**